bundles/org.openhab.binding.millheat/src/main/java/org/openhab/binding/millheat/internal/handler/MillheatAccountHandler.java

```diff
@@ -0,0 +1,449 @@
+import org.openhab.binding.millheat.internal.model.ModeType;
+import org.openhab.binding.millheat.internal.model.Room;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link MillheatAccountHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Arne Seime - Initial contribution
+ */
+public class MillheatAccountHandler extends BaseBridgeHandler {
+    private static final int MIN_TIME_BETWEEEN_MODEL_UPDATES_MS = 30_000;
+    private static final int NUM_NONCE_CHARS = 16;
+    private static final String CONTENT_TYPE = "application/x-zc-object";
+    public static String API_ENDPOINT_1 = "https://eurouter.ablecloud.cn:9005/zc-account/v1/";
+    public static String API_ENDPOINT_2 = "https://eurouter.ablecloud.cn:9005/millService/v1/";
+    private static final String ALLOWED_NONCE_CHARACTERS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+    private static final String REQUEST_TIMEOUT = "300";
+    private final Logger logger = LoggerFactory.getLogger(MillheatAccountHandler.class);
+    private String userId;
+    private String token;
+    private final HttpClient httpClient;
+    private RequestLogger requestLogger;
+    private MillheatDiscoveryService discoveryService;
+    private Gson gson;
+    private MillheatModel model = new MillheatModel(0);
+    private @Nullable ScheduledFuture<?> statusFuture;
+    private MillheatAccountConfiguration config;
+    private Map<ThingUID, ServiceRegistration<DiscoveryService>> discoveryServiceRegistrations = new HashMap<>();
+
+    private static String getRandomString(final int sizeOfRandomString) {
+        final Random random = new Random();
+        final StringBuilder sb = new StringBuilder(sizeOfRandomString);
+        for (int i = 0; i < sizeOfRandomString; ++i) {
+            sb.append(ALLOWED_NONCE_CHARACTERS.charAt(random.nextInt(ALLOWED_NONCE_CHARACTERS.length())));
+        }
+        return sb.toString();
+    }
+
+    public MillheatAccountHandler(Bridge bridge, HttpClient httpClient, BundleContext context) {
+        super(bridge);
+        this.httpClient = httpClient;
+        this.httpClient.getContentDecoderFactories().clear();
+        this.httpClient.setUserAgentField(new HttpField("User-Agent", "MillheatApp"));
+        BooleanSerializer serializer = new BooleanSerializer();
+
+        gson = new GsonBuilder().setPrettyPrinting().setDateFormat("yyyy-MM-dd HH:mm:ss")
+                .registerTypeAdapter(Boolean.class, serializer).registerTypeAdapter(boolean.class, serializer).create();
+
+        config = getConfigAs(MillheatAccountConfiguration.class);
+        discoveryService = new MillheatDiscoveryService(this);
+        ServiceRegistration<DiscoveryService> serviceRegistration = context.registerService(DiscoveryService.class,
```

Comment:
ServiceRegistration should be done in the Factory not in the constructor.
https://github.com/openhab/openhab2-addons/search?p=3&q=ServiceRegistration%3CDiscoveryService%3E&unscoped_q=ServiceRegistration%3CDiscoveryService%3E

There it can also be unregistered if the handler is removed.

Reply:
DOne